### PR TITLE
asdf-vm: init at 0.8.1

### PR DESCRIPTION
**Pull request by bolasblack**
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Added a release notes entry if the change is major or breaking
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/misc/asdf-vm/default.nix [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+    mkdir -p $out/etc/profile.d
+    echo "source $out/share/asdf-vm/asdf.sh" > $out/etc/profile.d/asdf.sh
+
+    mkdir -p $out/bin
+    makeWrapper $out/share/asdf-vm/bin/asdf $out/bin/asdf --set ASDF_DIR $out/share/asdf-vm
+
+    installShellCompletion --zsh --name _asdf completions/_asdf
+    installShellCompletion --fish --name asdf.fish completions/asdf.fish
+    installShellCompletion --bash --name asdf.bash completions/asdf.bash
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/asdf-vm/default.nix:37]:
```suggestion
    installShellCompletion --cmd asdf \
      --zsh completions/_asdf
      --fish completions/asdf.fish
      --bash completions/asdf.bash
```

File: pkgs/tools/misc/asdf-vm/default.nix [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  '';
+
+  meta = with lib; {
+    description = "Extendable version manager with support for Ruby, Node.js, Erlang & more";
+    homepage = "https://asdf-vm.com/";
+    license = licenses.mit;
+    maintainers = [ maintainers.c4605 ];
+    platforms = platforms.linux
+             ++ platforms.darwin;
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/asdf-vm/default.nix:46]:
```suggestion
    platforms = platforms.unix;
```

File: pkgs/tools/misc/asdf-vm/default.nix [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+    mkdir -p $out/bin
+    # asdf-vm create "shim" file like this:
+    #
+    #    exec $ASDF_DIR/bin/asdf exec "node" "$@"
+    #
+    # So we shouldn't set ASDF_DIR like $out/share/asdf-vm, because $out
+    # always change
+    makeWrapper $out/share/asdf-vm/bin/asdf $out/bin/asdf \
+      --run "export ASDF_DIR=\"\$HOME/.nix-profile/share/asdf-vm\""
```
Comment (bolasblack) [inline pkgs/tools/misc/asdf-vm/default.nix:40]: Hi @SuperSandro2000, do you think I resolve this problem in the right way?

**Comment by sebastiant** [comment]
Just wanted to chime in and say: Thanks for this, it's highly appreciated!

File: pkgs/tools/misc/asdf-vm/default.nix (hunk 3 above)
Comment (SuperSandro2000) [reply pkgs/tools/misc/asdf-vm/default.nix:40]: I am not sure. ``~/.nix-profile/share/asdf-vm`` is read-only on my system and is only created when installing the package, not when using it in nix-shell. We should probably hardcode this to the $out path.

```suggestion
      --set ASDF_DIR $out/share/asdf-vm
```

Comment (bolasblack) [reply pkgs/tools/misc/asdf-vm/default.nix]: I see, I just realized the code `echo "source $out/share/asdf-vm/asdf.sh" > $out/etc/profile.d/asdf.sh` will not work in nix-shell also

I tried a new idea (you can see my new temporary commit), but it's only working when upgraded asdf-vm and execute `asdf` command at first. If I use asdf-vm 0.7.8 installed nodejs, then upgrade asdf-vm, then execute `nix-collect-garbage`, then execute `node` immediately, shell will throw an error like `/nix/store/gmkxfwl44lxfr9li8bk9f0qdr5xz8vpv-asdf-vm-0.7.8/share/asdf-vm/bin/asdf: No such file or directory`

And also, I realized that `echo "source $out/share/asdf-vm/asdf.sh" > $out/etc/profile.d/asdf.sh` not working in nix-shell, but it is very important for asdf-vm 😂 

Do you have any good idea to resolve those problems?

**Comment by bolasblack** [comment]
Hi @SuperSandro2000, it's still not out-of-the-box in nix-shell (need `source $out/etc/profile.d/asdf-prepare.sh`), I really have no way :(

But I think it's ok now in normal install process

**Pull request merged**